hydroflow/examples/deadlock_detector/main.rs

```diff
@@ -0,0 +1,58 @@
+use clap::{ArgEnum, Parser};
+use hydroflow::tokio;
+use peer::run_detector;
+use serde::Deserialize;
+
+use std::error::Error;
+use std::fs::File;
+use std::io::BufReader;
+use std::path::Path;
+
+mod helpers;
+mod peer;
+mod protocol;
+
+/// This is a remedial distributed deadlock (cycle) detector
+#[derive(Clone, ArgEnum, Debug)]
+enum GraphType {
+    Mermaid,
+    Dot,
+    Json,
+}
+
+#[derive(Parser, Debug)]
+struct Opts {
+    #[clap(long)]
+    path: String,
+    #[clap(long)]
+    port: u16,
+    #[clap(long)]
+    addr: String,
+    #[clap(arg_enum, long)]
+    graph: Option<GraphType>,
+}
+
+#[derive(Deserialize, Debug)]
+struct Addresses {
+    peers: Vec<String>,
+}
+
+fn read_addresses_from_file<P: AsRef<Path>>(path: P) -> Result<Addresses, Box<dyn Error>> {
```

Comment:
(another thing that doesn't matter, but fun tip you can write `read(path: impl AsRef<Path>)` to avoid needing the explicit generic parameter)